### ISSUE TITLE
Empirical Fisher Estimation

### ISSUE THREAD

**Issue by tangbinh** [GMvandeVen/continual-learning#11]
It seems convenient to average the gradients over samples by calling `F.nll_loss` before squaring them, as we only need one backward pass. However, I feel like the diagonal of the empirical Fisher information matrix should be calculated by squaring the gradients before taking their average (as done in [this Tensorflow implementation](https://github.com/ariseff/overcoming-catastrophic/blob/master/model.py#L65-L68)). Can you please confirm that the order doesn't matter here?

My understanding is that the expected values of the gradients are 0 (see [this Wiki](https://en.wikipedia.org/wiki/Fisher_information)), so if you do averaging first, the Fisher values are very close to 0, which seems incorrect. Am I missing something here? Please let me know what you think. Thank you.

https://github.com/GMvandeVen/continual-learning/blob/ff0e03cb913ac0dea4fc59058968b1e6784decfd/continual_learner.py#L110-L125

**Comment by GMvandeVen** [comment]
Thanks for your comment! You’re right that the order in which the square and average operations are applied matters, exactly for the reason you point out. But the code in my repository does use the correct order, because the Fisher Information matrix is calculated with batches of size 1 ([see here](https://github.com/GMvandeVen/continual-learning/blob/ff0e03cb913ac0dea4fc59058968b1e6784decfd/continual_learner.py#L88-L92)).

I admit this might not be the most efficient implementation (alternative suggestions are very welcome!), but the reason I chose this implementation for now is that (as far as I’m aware) in PyTorch it is currently not possible to access the gradients of individual elements in a sum.

**Comment by tangbinh** [comment]
I see. Estimating the Fisher information matrix with a single example seems to result in high variance (see [this notebook](https://github.com/ariseff/overcoming-catastrophic/blob/master/experiment.ipynb) for example). Have you seen any difference if a larger batch size is used? As you said, I don't think it's possible to access individual gradients with just one backward call.

**Comment by GMvandeVen** [comment]
Sorry, I should have been clearer, I realise my use of ‘batches’ is a bit confusing here. What I meant is that the backward passes(*) are done one-by-one for each sample used the calculate the Fisher Information matrix. The number of samples used to calculate the Fisher Information matrix is typically not 1 (this is set by the option `--fisher-n`; the default is using the full training-set).

(*) actually also the forward passes; I now realise this could be made more efficient by at least performing the forward passes with larger ‘batches’. I’ll look into that when I get some time.
